Switch DirectoryControl to use AsnWriter, AsnDecoder

edwardneal opened the pull request:
Relates to #97540.

This PR replaces all references to `BerConverter` in LDAP directory control generation/parsing to use AsnWriter and AsnDecoder. SortRequestControl didn't use BerConverter directly - it called the OpenLDAP and WLDAP `ldap_create_sort_control` APIs instead. This class was the only thing in S.DS.P which referenced the `ldap_create_sort_control` and `SortKeyInterop` struct, so I deleted them both.

#### SortRequestControl
The change to SortRequestControl's generation mechanism might also resolve #34679, since there shouldn't be any mechanism for the heap corruption to occur.

Most of the SortRequestControl's new ASN.1 encoding is pretty uncontroversial, but there was a bit of discussion in PR #65548 around the encoding of the sort key's attribute name, and this was marshalled (as part of SortKeyInterop) with different encodings between Windows and Linux. In the RFC, this is defined (indirectly) as an LdapString; this is described as ISO10646 characters, encoded as a UTF-8 string and represented as an OCTET STRING. I'm fairly sure that UTF8Encoding.GetBytes fulfils this, and running the associated test case against a real AD domain controller passes.

#### Test changes
There are also test changes, but these are largely to change the special-casing of expected byte values between OpenLDAP and WLDAP - .NET now generates these values in a consistent format (the OpenLDAP format) regardless of platform. The .NET Framework tests continued to use the version of S.DS.P from the GAC in my environment, so I've special-cased by the framework version rather than by the platform.

#### Misc. optimizations
There were a handful of byte-by-byte array copies, which I've switched over to using span-based copies in hopes that they'll benefit slightly from vectorisation. TransformControls and GetValue have a related change: where they used to reference properties returning byte arrays (which took defensive copies) they now reference the property values directly. These should both reduce GC traffic slightly.

edwardneal committed:
Removed BerConverter.Decode usage

Replaced this with the managed AsnDecoder, removing PInvoke from a potential hot path.

edwardneal committed:
Removed BerConverter.Encode usage

Also removed the manual API calls to ldap_create_sort_control - this is now built in managed code.
This then has knock-on effects to eliminate the SortKeyInterop classes.

edwardneal committed:
Regression in CrossDomainMoveControl

edwardneal committed:
Correcting test values

Most of the Control tests were hardcoded to the output of BerConverter, which uses four-byte lengths in all cases.
This behaviour is now different: the same output is returned across all platforms for .NET, and remains unchanged for .NET Framework.
This should also close issue 34679.

edwardneal committed:
Removed SortKeyInterop from .csproj

edwardneal committed:
Small optimisations

Reduce number of copies required in TransformControls, and enable these copies to take advantage of newer intrinsics where available.

edwardneal committed:
AD compatibility test corrections

Windows domain controllers may return a distinguished name starting with OU=, rather than ou=.

PaulusParssinen commented:
Out of curiosity, any benchmarks for perf. numbers before/after switching to `AsnReader`/`AsnWriter`?

edwardneal commented:
I've not got benchmarks right now, but will write some in the next few days. In advance of these, I expect there'll be a modest reduction in managed and unmanaged memory usage, and that execution time will reduce (while remaining within the margin of error for the network request itself.)

edwardneal committed:
Performance improvement, bugfix

Preallocating space for AsnWriter buffers to reduce memory usage.
Correctly handling attribute names in SortControls.

edwardneal commented:
Benchmarks are below. To summarize:
* As expected, the performance improvements are working in the margins. I fully expect most of the execution time variations to be lost in the noise of network traffic.
* 35% median reduction in memory usage. One notable improvement on this in `DirectoryControl.TransformControls`, which is on the hot path for processing LDAP responses and reduces memory usage by about 75%.
* Although the percentage reductions in memory usage are good, the absolute reductions are pretty small - the median reduction was of 144 bytes.
* 88.5% median reduction in execution time, although the absolute reductions are often small - AsqRequest is reduced from 1.869 microseconds to 185.1 nanoseconds.
* `DirectoryControl.TransformControls` is another notable exception to this, reducing from 6.596us to 1.588us.
* Most of the original code's memory allocations stuck around for Gen1 GCs. This GC pressure no longer exists.
* I've got no data on unmanaged memory usage. This is particularly relevant for `SortRequest`, which moved from 400 bytes to 416 bytes managed memory usage. I'm assuming that this lack of data is the reason for the increase in memory usage - it's not actually increasing, it's just now trackable in the managed counters.
* Code size is 15 bytes in most places. The disassembly puts this at the size of the benchmark itself - just enough to return `DirectoryControl.GetValue`. I think this is just noise from the JIT inlining.

<details>
<summary>Performance header</summary>

```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3296/23H2/2023Update/SunValley3)
Intel Core i7-8565U CPU 1.80GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 8.0.200
  [Host]     : .NET 8.0.4 (8.0.424.16909), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.4 (8.0.424.16909), X64 RyuJIT AVX2
```
</details>

<details>
<summary>AsqRequestControl.GetValue: -90% execution time, -35% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.869 μs | 0.0373 μs | 0.0485 μs |     127 B | 0.1030 | 0.0992 |     432 B |
| PR | 185.1 ns | 3.25 ns | 4.12 ns | 15 B | 0.0668 | | 280 B |

</details>

<details>
<summary>CrossDomainMoveControl.GetValue: -54% execution time, -62% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 89.81 ns | 1.297 ns | 1.332 ns |     1,039 B | 0.0516 | |     216 B |
| PR | 41.24 ns | 0.469 ns | 0.438 ns | 2,577 B | 0.0191 | | 80 B |

</details>

<details>
<summary>DirSyncRequestControl.GetValue: -89% execution time, -35% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.587 μs | 0.0453 μs | 0.1329 μs |     191 B | 0.1221 | 0.1183 |     520 B |
| PR | 162.4 ns | 2.20 ns | 2.06 ns | 15 B | 0.0782 | | 328 B |

</details>

<details>
<summary>ExtendedDNControl.GetValue: -89% execution time, -30% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.059 μs | 0.0213 μs | 0.0522 μs |     147 B | 0.0877 | 0.0858 |     368 B |
| PR | 115.9 ns | 1.72 ns | 1.44 ns | 15 B | 0.0610 | | 256 B |

</details>

<details>
<summary>PageResultRequestControl.GetValue: -90% execution time, -40% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.412 μs | 0.0283 μs | 0.0432 μs |     160 B | 0.1030 | 0.1011 |     432 B |
| PR | 134.7 ns | 1.49 ns | 1.32 ns | 15 B | 0.0610 | | 256 B |

</details>

<details>
<summary>QuotaControl.GetValue: -92% execution time, -28% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.607 μs | 0.0261 μs | 0.0232 μs |     127 B | 0.0916 | 0.0877 |     392 B |
| PR | 120.4 ns | 1.65 ns | 1.29 ns | 15 B | 0.0668 | | 280 B |

</details>

<details>
<summary>SearchOptionsControl.GetValue: -90% execution time, -30% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.251 μs | 0.0223 μs | 0.0197 μs |     147 B | 0.0877 | 0.0858 |     368 B |
| PR | 114.7 ns | 1.56 ns | 1.39 ns | 15 B | 0.0610 | | 256 B |

</details>

<details>
<summary>SecurityDescriptorFlagControl.GetValue: -88% execution time, -30% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.021 μs | 0.0188 μs | 0.0460 μs |     147 B | 0.0877 | 0.0858 |     368 B |
| PR | 115.8 ns | 1.47 ns | 1.37 ns | 15 B | 0.0610 | | 256 B |

</details>

<details>
<summary>SortRequestControl.GetValue: -79% execution time, +4% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.893 μs | 0.0220 μs | 0.0195 μs |     15 B | 0.0954 | |     400 B |
| PR | 390.5 ns | 5.37 ns | 5.03 ns | 15 B | 0.0992 | | 416 B |

</details>

<details>
<summary>DirectoryControl.TransformControls: -75% execution time, -76% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 6.596 μs | 0.1310 μs | 0.3164 μs |     3,603 B | 0.8392 | 0.0153 |     3.43 KB |
| PR | 1.588 μs | 0.0208 μs | 0.0255 μs | 12,854 B | 0.1945 | | 816 B |

</details>

<details>
<summary>VerifyNameControl.GetValue: -87% execution time, -46% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.575 μs | 0.0480 μs | 0.1240 μs |     1,526 B | 0.1488 | 0.1469 |     624 B |
| PR | 195.1 ns | 1.43 ns | 1.27 ns | 15 B | 0.0801 | | 336 B |

</details>

<details>
<summary>VlvRequestControl.GetValue: -84% execution time, -69% Gen0 memory allocation</summary>

| Method    | Mean     | Error     | StdDev    | Code Size | Gen0   | Gen1   | Allocated |
|---------- |---------:|----------:|----------:|----------:|-------:|-------:|----------:|
| Original | 1.459 μs | 0.0276 μs | 0.0650 μs |     15 B | 0.2289 | 0.0038 |     968 B |
| PR | 227.7 ns | 3.21 ns | 3.00 ns | 15 B | 0.0706 | | 296 B |

</details>

I've made performance adjustment by specifying the initial size of AsnWriter, since this is trivial to calculate (or always static.) AsnWriter grows in 1KB increments, which is much larger than the size of a normal directory control and causes memory usage to balloon.

One inefficiency which I couldn't eliminate is that when writing strings as ASN.1 octet strings, I want to manually select the encoding to use and encode directly into the AsnWriter buffer. This isn't possible, (probably to keep AsnWriter specification-compliant) so I have to reserve/allocate a byte array, encode into that and write that out as an octet string. An example of this behaviour is in `VerifyNameControl.GetValue`.

PaulusParssinen commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -432,14 +515,33 @@ public string ServerName
+
+                if (!string.IsNullOrEmpty(ServerName))
+                {
+                    int serverNameLength = Encoding.Unicode.GetByteCount(ServerName);
+                    Span<byte> tmpValue = serverNameLength < 256 ? stackalloc byte[serverNameLength] : new byte[serverNameLength];
```
nit:
```suggestion
                    int serverNameLength = Encoding.Unicode.GetByteCount(ServerName);
                    Span<byte> tmpValue = serverNameLength <= 256 ? stackalloc byte[serverNameLength] : new byte[serverNameLength];
```

also I wonder if this is a place for constant sized stackalloc? (IIRC it has bit better codegen)
e.g. 
```cs
const int MaxStackalloc = 256;
.. = serverNameLength <= MaxStackalloc ? 
    stackalloc byte[MaxStackalloc].Slice(0, serverNameLength) : new byte[serverNameLength];
```

PaulusParssinen commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,99 +126,128 @@ public virtual byte[] GetValue()
-                        // decoding might fail as attribute is optional
-                        o = BerConverter.Decode("{e}", value);
-                        Debug.Assert(o != null && o.Length == 1);
+                        Span<byte> attributeNameBuffer = octetStringLength < scratchSpace.Length ? scratchSpace.Slice(0, octetStringLength) : new byte[octetStringLength];
```
nit: use all the bytes available in the scratch buffer
```suggestion
                        Span<byte> attributeNameBuffer = octetStringLength <= scratchSpace.Length ? scratchSpace.Slice(0, octetStringLength) : new byte[octetStringLength];
```

edwardneal commented:
Thanks @PaulusParssinen, I'll correct the nits shortly.

I dug into the codegen for stackalloc. The results were fairly interesting.

<details>
<summary>Variable string length stackalloc</summary>

#### Sample code
```csharp
Span<byte> byteSpan = stackalloc byte[myString.Length];
return byteSpan.Length;
```

#### Assembly
```
7FFACF4D6C90 push      rbp
7FFACF4D6C91 sub       rsp,30
7FFACF4D6C95 lea       rbp,[rsp+20]
7FFACF4D6C9A mov       rax,613C5BC56FC9
7FFACF4D6CA4 mov       [rbp+8],rax
7FFACF4D6CA8 mov       rax,[rcx+8]
7FFACF4D6CAC mov       eax,[rax+8]
7FFACF4D6CAF mov       ecx,eax
7FFACF4D6CB1 test      rcx,rcx
7FFACF4D6CB4 je        short 00007FFACF4D6CD4
7FFACF4D6CB6 add       rcx,0F
7FFACF4D6CBA shr       rcx,4
7FFACF4D6CBE add       rsp,20
7FFACF4D6CC2 push      0
7FFACF4D6CC4 push      0
7FFACF4D6CC6 dec       rcx
7FFACF4D6CC9 jne       short 00007FFACF4D6CC2
7FFACF4D6CCB sub       rsp,20
7FFACF4D6CCF lea       rcx,[rsp+20]
7FFACF4D6CD4 mov       rcx,613C5BC56FC9
7FFACF4D6CDE cmp       [rbp+8],rcx
7FFACF4D6CE2 je        short 00007FFACF4D6CE9
7FFACF4D6CE4 call      CORINFO_HELP_FAIL_FAST
7FFACF4D6CE9 nop
7FFACF4D6CEA lea       rsp,[rbp+10]
7FFACF4D6CEE pop       rbp
7FFACF4D6CEF ret
```
</details>

<details>
<summary>Constant length stackalloc</summary>

#### Sample code
```csharp
Span<byte> byteSpan = stackalloc byte[0x100];
return byteSpan.Length;
```

#### Assembly
```
7FFB47536CD0 push      rbp
7FFB47536CD1 sub       rsp,30
7FFB47536CD5 lea       rbp,[rsp+20]
7FFB47536CDA mov       rax,2DEAD4CD135B
7FFB47536CE4 mov       [rbp+8],rax
7FFB47536CE8 mov       eax,100
7FFB47536CED test      rax,rax
7FFB47536CF0 je        short 00007FFB47536D10
7FFB47536CF2 add       rax,0F
7FFB47536CF6 shr       rax,4
7FFB47536CFA add       rsp,20
7FFB47536CFE push      0
7FFB47536D00 push      0
7FFB47536D02 dec       rax
7FFB47536D05 jne       short 00007FFB47536CFE
7FFB47536D07 sub       rsp,20
7FFB47536D0B lea       rax,[rsp+20]
7FFB47536D10 mov       eax,100
7FFB47536D15 mov       rcx,2DEAD4CD135B
7FFB47536D1F cmp       [rbp+8],rcx
7FFB47536D23 je        short 00007FFB47536D2A
7FFB47536D25 call      CORINFO_HELP_FAIL_FAST
7FFB47536D2A nop
7FFB47536D2B lea       rsp,[rbp+10]
7FFB47536D2F pop       rbp
7FFB47536D30 ret
```
</details>

<details>
<summary>Constant length stackalloc, with SkipLocalsInit</summary>

#### Sample code
```csharp
Span<byte> byteSpan = stackalloc byte[0x100];
return byteSpan.Length;
```

#### Assembly
```
7FFACF4D6C50 push      rbp
7FFACF4D6C51 sub       rsp,30
7FFACF4D6C55 lea       rbp,[rsp+20]
7FFACF4D6C5A mov       rax,0D5EB80FED67D
7FFACF4D6C64 mov       [rbp+8],rax
7FFACF4D6C68 test      [rsp],esp
7FFACF4D6C6B sub       rsp,100
7FFACF4D6C72 lea       rax,[rsp+20]
7FFACF4D6C77 mov       eax,100
7FFACF4D6C7C mov       rcx,0D5EB80FED67D
7FFACF4D6C86 cmp       [rbp+8],rcx
7FFACF4D6C8A je        short 00007FFACF4D6C91
7FFACF4D6C8C call      CORINFO_HELP_FAIL_FAST
7FFACF4D6C91 nop
7FFACF4D6C92 lea       rsp,[rbp+10]
7FFACF4D6C96 pop       rbp
7FFACF4D6C97 ret
```
</details>

The general pattern of the assembly is thus:
```
; Boilerplate
7FFB47536CD0 push      rbp
7FFB47536CD1 sub       rsp,30
7FFB47536CD5 lea       rbp,[rsp+20]
7FFB47536CDA mov       rax,2DEAD4CD135B
7FFB47536CE4 mov       [rbp+8],rax

; If the localsinit flag is set:
    ; eax / rax holds the value to expand the stack by.
    ; If the value is constant:
        7FFB47536CE8 mov       eax,100
        7FFB47536CED test      rax,rax
    ; If the value is dynamic:
        7FFB47536CA8 mov       rax,[rcx+8]
        7FFB47536CAC mov       eax,[rax+8]
        7FFB47536CAF mov       ecx,eax
        7FFB47536CB1 test      rcx,rcx

    ; Boilerplate - stack smashing checking
    7FFB47536CF0 je        short 00007FFB47536D10

    ; Relevant section.
    7FFB47536CF2 add       rax,0F
    7FFB47536CF6 shr       rax,4
    7FFB47536CFA add       rsp,20
    7FFB47536CFE push      0
    7FFB47536D00 push      0
    7FFB47536D02 dec       rax
    7FFB47536D05 jne       short 00007FFB47536CFE

    7FFB47536D07 sub       rsp,20

; If the localsinit flag isn't set, we just subtract from rsp directly:
    7FFACF4D6C68 test      [rsp],esp
    7FFACF4D6C6B sub       rsp,100

; Boilerplate (including the return value)
7FFB47536D0B lea       rax,[rsp+20]
7FFB47536D10 mov       eax,100
7FFB47536D15 mov       rcx,2DEAD4CD135B
7FFB47536D1F cmp       [rbp+8],rcx
7FFB47536D23 je        short 00007FFB47536D2A
7FFB47536D25 call      CORINFO_HELP_FAIL_FAST
7FFB47536D2A nop
7FFB47536D2B lea       rsp,[rbp+10]
7FFB47536D2F pop       rbp
7FFB47536D30 ret
```

The relevant section is surprising. localsinit affects over a stackalloc'd Span, even though the documentation is [fairly specific](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/stackalloc) in saying that this might not happen:
> The content of the newly allocated memory is undefined. You should initialize it before the use. For example, you can use the [Span<T>.Clear](https://learn.microsoft.com/en-us/dotnet/api/system.span-1.clear) method that sets all the items to the default value of type T.

I'll mark the DirectoryControls method to disable localsinit, then make the Span length constant. If localsinit were to be left enabled, without any intrinsics helping the zeroing along, switching to using a constant 256-byte Span is slower than a dynamic (but shorter) one. The benchmark for this is:

| Method                           | Mean     | Error     | StdDev    | Ratio | RatioSD | Code Size | 
|--------------------------------- |---------:|----------:|----------:|------:|--------:|----------:|
| DynamicStackAlloc                         | 8.106 ns | 0.1953 ns | 0.2171 ns |  1.00 |    0.00 |      96 B |
| DynamicStackAllocSkipLocalsInit  | 1.345 ns | 0.0419 ns | 0.0392 ns |  0.17 |    0.01 |     121 B |
| ConstantStackAlloc            | 9.204 ns | 0.1208 ns | 0.1071 ns |  1.13 |    0.04 |      97 B |
| ConstantStackAllocSkipLocalsInit | 1.138 ns | 0.0274 ns | 0.0214 ns |  0.14 |    0.00 |      72 B |


It doesn't change the control-by-control benchmarks earlier in the PR though - it's a rounding error in comparison.

PaulusParssinen commented:
IIRC, BCL is [SkipLocalsInit] already.. or is that limited to certain assemblies, not sure.

edwardneal committed:
Following code review

Corrected off-by-one error in stack allocation comparison.
Made some stackallocs constant (unless a meaningful constant size would be >256 bytes) and reviewed these constant values based on real-world attribute lengths/server names.

edwardneal commented:
Excellent, thanks. I've fixed the off-by-one error and made all but one of the stackallocs constant. In the case of attribute names, I've then cut the threshold from 256 -> 64, since the largest AD attribute name is 53 bytes.

The stackalloc which I've kept dynamic is the server name in VerifyNameControl. Looking at RFC1035, a three-label FQDN could be up to 384 bytes when UTF-16 encoded - potentially 512 bytes in the case of an enterprise AD environment. I think that at that point, the elimination of two MOVs is outweighed by the cost of always allocating 512 bytes of stack space.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
+                    // The control is a PageControl, as described in RFC 2696.
+                    byte[] cookie = Array.Empty<byte>();
+
+                    AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
```
AsnDecoder is going to throw AsnContentException if anything goes wrong.  BerConverter would have thrown BerConversionException.

I don't see any try/catch that would normalize this.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
+                    // The control is a PageControl, as described in RFC 2696.
+                    byte[] cookie = Array.Empty<byte>();
+
+                    AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
```
Assuming that there's not supposed to be any surplus data, you should read the `bytesConsumed` value and throw if it's not `asnSpan.Length`

edwardneal replied:
`BerConverter.Decode` doesn't throw an exception if there's anything after the end of the sequence, so I'll maintain compatibility with that.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
-                    int size = (int)result[0];
-                    // user expects cookie with length 0 as paged search is done.
-                    byte[] cookie = (byte[])result[1] ?? Array.Empty<byte>();
+                    asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan.Slice(sequenceContentOffset), AsnEncodingRules.BER, out int size, out int bytesConsumed);
```
AsnDecoder.ReadSequence doesn't check that the contents don't "reach out" (particularly if a definite length encoding is used).  So, if for some reason it's not viable to check that the sequence was the full byte range earlier, this needs to use a sliced down input.  `asnSpan.Slice(sequenceContentOffset, sequenceContentLength)` being the most obvious, but that makes things harder for the next call; so really you'd just do

```suggestion
                    asnSpan = asnSpan.Slice(sequenceContentOffset, sequenceContentLength);
                    asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int size, out int bytesConsumed);
```

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
-                    // user expects cookie with length 0 as paged search is done.
-                    byte[] cookie = (byte[])result[1] ?? Array.Empty<byte>();
+                    asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan.Slice(sequenceContentOffset), AsnEncodingRules.BER, out int size, out int bytesConsumed);
+                    Debug.Assert(asnReadSuccessful);
```
What if the sender sends non-compliant data?

Instead of an assert, this should be

```c#
if (!asnReadSuccessful)
{
    throw new BerConversionException();
}
```

or something like that.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
+
+                        // user expects cookie with length 0 as paged search is done.
+                        asnReadSuccessful = AsnDecoder.TryReadOctetString(asnSpan, cookie, AsnEncodingRules.BER, out _, out _);
+                        Debug.Assert(asnReadSuccessful);
```
-assert +if-false-throw.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
+                        // user expects cookie with length 0 as paged search is done.
+                        asnReadSuccessful = AsnDecoder.TryReadOctetString(asnSpan, cookie, AsnEncodingRules.BER, out _, out _);
+                        Debug.Assert(asnReadSuccessful);
+                    }
```
```suggestion
                    }
                    else
                    {
                        throw ...
                    }
```

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
-                    Debug.Assert((o != null) && (o.Length == 1));
+                    // The control is an AsqControl, as described in MS-ADTS section 3.1.1.3.4.1.18
+                    AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out _);
+                    Debug.Assert(sequenceContentLength > 0);
```
The same feedback from the previous section will apply to all of these object reads.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,99 +137,128 @@ public virtual byte[] GetValue()
+                    asnReadSuccessful = AsnDecoder.TryReadEncodedValue(asnSpan, AsnEncodingRules.BER, out Asn1Tag octetStringTag, out _, out int octetStringLength, out _);
+                    if (asnReadSuccessful)
                     {
-                        // decoding might fail as attribute is optional
-                        o = BerConverter.Decode("{e}", value);
-                        Debug.Assert(o != null && o.Length == 1);
+                        Span<byte> attributeNameBuffer = octetStringLength <= AttributeNameStackAllocationThreshold ? attributeNameScratchSpace.Slice(0, octetStringLength) : new byte[octetStringLength];
 
-                        result = (int)o[0];
+                        _ = AsnDecoder.TryReadOctetString(asnSpan, attributeNameBuffer, AsnEncodingRules.BER, out _, out _);
+                        attribute = s_utf8Encoding.GetString(attributeNameBuffer);
                     }
```
TryReadEncodedValue can fail for a lot of sub-reasons; but I _think_ you're just using it for "there was still data".  In which case you should rewrite this to

```C#
if (!asnSpan.IsEmpty)
{
    scoped ReadOnlySpan<byte> attributeNameBuffer;

    if (asnSpan.Length <= AttributeNameStackAllocationThreshold)
    {
        asnReadSuccessful = AsnDecoder.TryReadOctetString(
            asnSpan,
            attributeNameScratchSpace,
            AsnEncodingRules.BER,
            out int consumed,
            out int written);

        // TryReadOctetString can't fail when the output buffer is at least as big as the input buffer.
        Debug.Assert(asnReadSuccessful);
        attributeNameBuffer = attributeNameScratchSpace.Slice(0, written);
        asnSpan = asnSpan.Slice(consumed);
    }
    else
    {
        attributeNameBuffer = AsnDecoder.ReadOctetString(
            asnSpan,
            AsnEncodingRules.BER,
            out int consumed);

        asnSpan = asnSpan.Slice(consumed);
    }

    attribute = s_utf8Encoding.GetString(attributeNameBuffer);
}

if (!asnSpan.IsEmpty)
{
    throw new BerConversionException();
}

SortResponseControl sort = ...;
```

bartonjs replied:
If this library wants to change to using array pool for "too big for the stack", then you'd go with something like your current code.  But once you're using `new byte[len]` you should just call the array-allocating variety.

edwardneal replied:
I'm using the pattern of `TryReadEncodedValue` + `TryReadOctetString` to do a few things. The octet strings in question are optional, so I'm interrogating the buffer to figure out it's about to yield an octet string and if so, how big it is. I'm also adding a post-hoc check to make sure that it's the last thing in the sequence, in line with the prior review comment.

bartonjs replied:
Let's assume you have a definite-length payload: `30 04 0A 01 01 04`.

* ReadSequence will report success.  The definite length encoding means it isn't going to bother checking that the children make sense.
* ReadEnumerated will consume 3 bytes (`0A 01 01`).  The sequence contents now have one byte left (which is never valid).
* TryReadEncodedValue will return false.  It can read a tag, but not a length, you're out of data.  Nothing was _corrupt_, so no exception, just a false return.
* What state are you in now?

The code I wrote is "there's more stuff.  Just read it as an OCTET STRING, because that's the only optional thing remaining".  That will cause an exception, because the data is insufficient and therefore corrupt.  If there was NO data remaining, it'd not read anything else, and satisfy the OPTIONAL.

I don't know what BerConverter would do in this invalid-data case.  If it previously rejected it (which I hope it would) then it needs to continue to be rejected.  If it's super loosey goosey, then what you have is more compatible (though I think the more strict read is still better)

edwardneal replied:
Your example SortControl payload and explanation makes sense. Incidentally, BerConverter will reject that if I use the `{eO}` format specifier.

That payload should naturally fail for either of two reasons:
* An octet string is expected to follow the enumerated value, and the data which actually follows that enumerated value isn't a valid octet string.
* The data following the enumerated value isn't an octet string, so we're expecting the enumerated value to be the last thing in the sequence. It isn't.

It doesn't right now because TryReadEncodedValue will return false and the updated code will ignore the extra data inside the payload's sequence. I need to fix that when handling PageControl, SortControl and VlvResponseControl.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -389,7 +472,19 @@ public SearchOption SearchOption
+             * SearchOptionsRequestValue ::= SEQUENCE {
+             *                                  Flags     INTEGER }
+             */
+            using (AsnWriter.Scope outerSequence = writer.PushSequence())
```
We used a nested type here because the expected code is just

```suggestion
            using (writer.PushSequence())
```

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -389,7 +472,19 @@ public SearchOption SearchOption
         {
-            _directoryControlValue = BerConverter.Encode("{i}", new object[] { (int)SearchOption });
+            int sizeEstimate = 8;
+            AsnWriter writer = new(AsnEncodingRules.BER, sizeEstimate);
```
If you're trying to minimize overall allocation here, you can pool your writers.  Even if it's just

```C#
[ThreadStatic]
private static AsnWriter t_sharedWriter;
```

Whether you do one per type, or a central one, or whatever.  You could also do more complicated pooling, of course.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -699,101 +826,71 @@ public SortKey[] SortKeys
                 for (int i = 0; i < value.Length; i++)
                 {
                     _keys[i] = new SortKey(value[i].AttributeName, value[i].MatchingRule, value[i].ReverseOrder);
+                    _keysAsnLength += 13 + (value[i].AttributeName?.Length ?? 0) + (value[i].MatchingRule?.Length ?? 0);
```
I strongly advise ditching all of the size estimation and just using a shared writer with the default 1k first page.

edwardneal commented:
Thanks @bartonjs - your review makes sense, and I'll apply the feedback later today.

What's your reasoning behind ditching the size estimation? It seems to be a fairly natural progression to shrink the first allocated block to a more appropriate size (since most controls will never even come close to 1KB in size.) From there, the sizes are either constant (because the control's structure is a set of fixed-size fields) or trivially calculated; in both cases, the length of each control is implied by the specification which defines it, so I don't see it as violating separation of concerns. AsnWriter will expand if the estimate is wrong, so I don't think it's creating any exceptions in the edge case. Is it a problem solely because the estimation makes it difficult to pool the AsnWriter instances, or does it have a problem in its own right?

bartonjs commented:
> What's your reasoning behind ditching the size estimation?

Assuming the objects are pooled centrally, it's that a low estimate from one area just forces a resize to happen somewhere else.  And otherwise that it's a lot of fiddly calculations to explain, especially in the ones that aren't just a sequence of an integer.

If they're pooled by class that needs them, then I suppose the "this is the size I should need here" applies... at least for rigidly sized ones.

edwardneal committed:
Implemented further code review comments

* Added limited caching of AsnWriters, as small/medium/large buckets.
* Replaced several Debug.Asserts with BerExceptions.
* Changed the type of raised exception to retain backwards-compatibility.
* Added extra checks on BER content lengths to match BerConverter.

edwardneal commented:
That makes sense, thanks. I've changed this to use a set of up to three AsnWriters based on expected payload length (0-32, 33-128, 129+) and this seems to work well. As a result, everything except for `AsqRequestControl`, `VerifyNameControl` and `SortRequestControl` can use the smallest pooled buffer. I expect that most of the time we'll see the three remaining controls use the middle pooled buffer, but I've added the unbounded one there as a fallback with a more appropriate initial size of 256 bytes.

The changes have slightly improved performance, but nothing to write home about - about 10-20ns shaved off the previous changes.

The encode/decode tests pass locally, but I've not currently got access to an AD server to perform real-world testing.

edwardneal committed:
Remainder of code review feedback

DirectoryControl payload parsing is stricter, throwing exception when invalid sequences are discovered instead of optional OCTET STRINGs.

edwardneal commented:
I've run the usual LDAP tests against a Windows domain controller, and all have passed as expected (including `SortRequestControl`, which was the only DirectoryControl left untested after the AsnWriter pooling changes.) Sorry about the wait!

I'm happy with the changes made to pool AsnWriters. With the inlining hint on `DirectoryControl.GetWriter`, I think this should JIT back down to a constant field access for cases such as `ExtendedDNControl` which have a constant size.

ericstj commented:
@buyaa-n @bartonjs can you have another look here?

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
+                for (int i = 0; i < controls.Length; i++)
                 {
-                    // The control is a PageControl.
-                    object[] result = BerConverter.Decode("{iO}", value);
```
While looking to see if there are any interesting edge cases popping up from this, I found a case where we're increasing the range of supported values, in a surprising way.

The DER sequence `30 05 02 01 03 04 00` is a `SEQUENCE(INTEGER(3), OCTET_STRING(''))`.  Under BER (and CER) it can (or, for CER, must) also be encoded with an indefinite length: `30 80 02 01 03 04 00 00 00` (change the length of the sequence from 0x05 to 0x80 and add two 0x00 values on the end)

I was incredibly surprised to learn that `BerConverter.Decode` doesn't seem to support indefinite-length encoding.

If it's important to continue NOT supporting indefinite-length encoding, that will require special handling.

The main thing I was trying to test for is "does the existing support trailing data", since the new code seems to.  And it looks like yes (as measured from the perspective of BerDecoder).

edwardneal replied:
That's caught me by surprise too, thanks for pointing it out. `BerConverter.Decode`'s lack of support for indefinite-length encoding is consistent between Windows and Linux - they may have interpreted [section 5.1 of RFC4511](https://datatracker.ietf.org/doc/html/rfc4511#section-5.1), which states that LDAP shall be encoded for exchange using BER and only use the definite form of length encoding, in order to ease the encoding/decoding overhead. Since their `ber_scanf` interpretation knows it'll never need to deal with indefinite length encoding, it's just not supported.

I don't think it precludes us from using an ASN.1 parser which supports it. Anyone who sends this is breaching the protocol specification. Section 4.1.1 notes:
> If the server receives an LDAPMessage from the client in which the
> LDAPMessage SEQUENCE tag cannot be recognized, the messageID cannot
> be parsed, the tag of the protocolOp is not recognized as a request,
> or the encoding structures or lengths of data fields are found to be
> incorrect, then the server SHOULD return the Notice of Disconnection
> described in [Section 4.4.1](https://datatracker.ietf.org/doc/html/rfc4511#section-4.4.1), with the resultCode set to protocolError,
> and MUST immediately terminate the LDAP session as described in
> [Section 5.3](https://datatracker.ietf.org/doc/html/rfc4511#section-5.3).
>
> In other cases where the client or server cannot parse an LDAP PDU,
> it SHOULD abruptly terminate the LDAP session ([Section 5.3](https://datatracker.ietf.org/doc/html/rfc4511#section-5.3)) where
> further communication (including providing notice) would be
> pernicious.  Otherwise, server implementations MUST return an
> appropriate response to the request, with the resultCode set to
> protocolError.

As I read it: if the client receives an LDAP PDU containing an indefinite-length BER-encoded value from the server but can still parse this PDU without any errors, the LDAP specification doesn't explicitly define the expected client-side action.

I've read the intent of the RFC's restriction as being solely to reduce encoding/decoding overhead, so unless you'd prefer to flag the server's breach of RFC4511 with an exception, I'm comfortable with `DirectoryControl.TransformControls` silently parsing any indefinite-length BER value properly and proceeding as it currently does.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
-                    Debug.Assert((o != null) && (o.Length == 1));
+                    if (controls[i].Type == "1.2.840.113556.1.4.319")
+                    {
+                        // The control is a PageControl, as described in RFC 2696.
```
RFC 2696 does not describe a `PageControl` structure (that sequence of characters does not appear in the RFC).

This looks like it might be a `realSearchControlValue` from https://datatracker.ietf.org/doc/html/rfc2696#section-2.

edwardneal replied:
Yes, agreed. The comments are supposed to refer to the DirectoryControl-derivatives being constructed, but this isn't accurate across the board. I'll change them and include the name of the structure they're being parsed from for clarity.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
-                    int result;
-                    string attribute = null;
-                    object[] o = BerConverter.TryDecode("{ea}", value, out bool decodeSucceeded);
+                        // If present, the remaining bytes in the control are expected to be an octet string.
+                        if (!asnSpan.IsEmpty)
```
This looks like it's declaring the `cookie` value to be OPTIONAL; but RFC 2696 does not describe it as being OPTIONAL.

The existing BerConverter.Decode also doesn't seem to support it being optional:

```C#
BerConverter.Decode("{iO}", new byte[] { 0x30, 0x05, 0x02, 0x01, 0x03, 0x04, 0x00 });
```

succeeds, but

```C#
BerConverter.Decode("{iO}", new byte[] { 0x30, 0x03, 0x02, 0x01, 0x03 });
```

(omitting the empty OCTET STRING value) throws a `BerConversionException`.

Since neither the existing code nor the referenced RFC mention this value as being optional, it should not be treated as such here.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
+                            cookie = AsnDecoder.ReadOctetString(asnSpan, AsnEncodingRules.BER, out bytesConsumed);
+                            asnSpan = asnSpan.Slice(bytesConsumed);
+                        }
+                        ThrowUnless(asnSpan.IsEmpty);
```
I'm a fan of this check (it matches what the RFC says), but it's not something that BerConverter.Decode does, so it's introducing an exception where there wasn't one previously.

That's not forbidden, but there should be a test for it (maybe there already is) and it needs to be part of a breaking change doc.

edwardneal replied:
There doesn't seem to be any self-contained test coverage of the controls decoded in this method; the closest it could be said to come is the processing which takes place when somebody tests against a real LDAP server. I'll add this coverage, then broaden it to explicitly test trailing data in SEQUENCEs.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
-                        count = (int)o[1];
-                        result = (int)o[2];
-                        context = (byte[])o[3];
+                        // The control is a DirSyncControl, as described in MS-ADTS section 3.1.1.3.4.1.3
```
`DirSyncControl` does not appear in that document (in revision 60).  Do you mean `DirSyncRequestValue`?

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
+                        ThrowUnless(sequenceContentLength > 0);
+                        asnSpan = asnSpan.Slice(sequenceContentOffset, sequenceContentLength);
+
+                        asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int moreData, out int bytesConsumed);
```
I believe `moreData` here corresponds to `Flags` from the DirSyncRequestValue structure, and should probably be named similarly.

bartonjs replied:
Technically, nothing constrains `Flags` to be System.Int32; but it looks like BerConverter.Decode `i` is treated as System.Int32; so this is just maintaining the limits that the previous version had.

If it just happened once or twice I'd give a comment saying why the code thought 32-bit was sufficient; but since it's pervasive... I'd consider putting a descriptive comment at the top, like

```
// Throughout this code, structure reads treat BER INTEGER values as System.Int32.
// Previous versions of this class, including the version in .NET Framework,
// used BerConverter.Decode with 'i' for the integer types, and that data type code is
// mapped to System.Int32.
// Therefore, when using AsnDecoder we have historical precedent to justify limiting to 32-bit.
```

edwardneal replied:
I've added the comments; thanks.
This control in specific corresponds to `DirSyncResponseValue`. `moreData` corresponds to the `MoreResults` field in that structure. I've renamed the variable to match it more closely though.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
+                        ThrowUnless(asnReadSuccessful);
+                        asnSpan = asnSpan.Slice(bytesConsumed);
+
+                        asnReadSuccessful = AsnDecoder.TryReadInt32(asnSpan, AsnEncodingRules.BER, out int count, out bytesConsumed);
```
The referenced spec calls this `MaxBytes`, and says that any value less than `0x100000` should be treated as `0x100000`.  I don't see that minimum-inflation here, nor in the constructor, so I assume it's a caller-interpretation problem.

This isn't changing any code flows around, so nothing _really_ needs to be done about it, but since I read it while wondering if Int32 was a valid assumption I dug a little deeper and was surprised.

edwardneal replied:
Since we're referring to a `DirSyncResponseValue`, the actual BER field name is `unused`. This doesn't have the same minimum-inflation logic behind it, but does raise the question of why exactly we read a field which the spec defines as being unused and treat it as a count. I don't think there's any action to be taken, but it's a little odd.

edwardneal committed:
First response to code review feedback

* Ensured that all references to DirectoryControl names in comments are accurate.
* Specified the name of the ASN.1 structure which is being parsed in the comments.
* Added explanatory comment stating that BER INTEGER values are being interpreted as 32-bit signed integers.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +159,163 @@ public virtual byte[] GetValue()
+                            }
+                            asnSpan = asnSpan.Slice(bytesConsumed);
+
+                            attribute = s_utf8Encoding.GetString(attributeNameBuffer);
```
Since this is using the throw-on-invalid form of the UTF-8 decoder there should be a catch for DecoderFallbackException (probably turning it into a BerConversionException)

bartonjs replied:
I think this is also a behavioral change which is virtuous but probably belongs in a breaking change doc.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +159,163 @@ public virtual byte[] GetValue()
+                        if (!asnSpan.IsEmpty)
+                        {
+                            // Attribute name is optional: AD for example never returns attribute name
+                            scoped Span<byte> attributeNameBuffer;
+
+                            if (asnSpan.Length <= AttributeNameStackAllocationThreshold)
+                            {
+                                asnReadSuccessful = AsnDecoder.TryReadOctetString(asnSpan, attributeNameScratchSpace, AsnEncodingRules.BER, out bytesConsumed, out int octetStringLength);
+                                Debug.Assert(asnReadSuccessful);
+                                attributeNameBuffer = attributeNameScratchSpace.Slice(0, octetStringLength);
+                            }
+                            else
+                            {
+                                attributeNameBuffer = AsnDecoder.ReadOctetString(asnSpan, AsnEncodingRules.BER, out bytesConsumed);
+                            }
+                            asnSpan = asnSpan.Slice(bytesConsumed);
+
+                            attribute = s_utf8Encoding.GetString(attributeNameBuffer);
```
Consider moving this to a helper routine... especially if it appears more than once.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +159,163 @@ public virtual byte[] GetValue()
+
+                            if (asnSpan.Length <= AttributeNameStackAllocationThreshold)
+                            {
+                                asnReadSuccessful = AsnDecoder.TryReadOctetString(asnSpan, attributeNameScratchSpace, AsnEncodingRules.BER, out bytesConsumed, out int octetStringLength);
```
"Surely you don't want this, you want ReadCharacterString(UTF8String)".

Nope, LDAPString is stupid.  `04` instead of `0C`.

bartonjs replied:
Assuming that this module is using `IMPLICIT TAGS`, I think that this needs to specify `new AsnTag(TagClass.ContextSpecific, 0)` for the expected tag here.

```asn
      SortResult ::= SEQUENCE {
         sortResult  ENUMERATED {
             success                   (0), -- results are sorted
             operationsError           (1), -- server internal failure
             timeLimitExceeded         (3), -- timelimit reached before
                                            -- sorting was completed
             strongAuthRequired        (8), -- refused to return sorted
                                            -- results via insecure
                                            -- protocol
             adminLimitExceeded       (11), -- too many matching entries
                                            -- for the server to sort
             noSuchAttribute          (16), -- unrecognized attribute
                                            -- type in sort key
             inappropriateMatching    (18), -- unrecognized or
                                            -- inappropriate matching
                                            -- rule in sort key
             insufficientAccessRights (50), -- refused to return sorted
                                            -- results to this client
             busy                     (51), -- too busy to process
             unwillingToPerform       (53), -- unable to sort
             other                    (80)
             },
       attributeType [0] AttributeDescription OPTIONAL }
```

If the module is `EXPLICIT TAGS` then the previous code was also wrong, as that `[0]` would mean it needed to be wrapped in a SEQUENCE.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -87,10 +89,44 @@ public bool ReverseOrder
+
         internal byte[] _directoryControlValue;
 
+        [ThreadStatic]
+        private static AsnWriter t_smallWriter;
+        [ThreadStatic]
+        private static AsnWriter t_mediumWriter;
+        [ThreadStatic]
+        private static AsnWriter t_largeWriter;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static AsnWriter GetWriter(int expectedSize)
+            => expectedSize switch
+            {
+                > 0 and <= 32 => t_smallWriter ??= new AsnWriter(AsnEncodingRules.BER, 32),
+                > 32 and <= 128 => t_mediumWriter ??= new AsnWriter(AsnEncodingRules.BER, 128),
+                _ => t_largeWriter ??= new AsnWriter(AsnEncodingRules.BER, 256)
+            };
```
```suggestion
        [ThreadStatic]
        private static AsnWriter t_writer;

        internal static AsnWriter GetWriter(int expectedSize)
            => t_writer ??= new AsnWriter(AsnEncodingRules.BER);
```

Basically, once two of these get activated, or one of them grows, you're better off just having one.

Only ever needed to write a 5 byte value `30 03 02 01 00` (`SEQUENCE(INTEGER(0))`)? Sure, 1019 bytes of the heap could _maybe_ be better used; but at the end of the day it doesn't really matter, especially once it's hanging out in a thread-static.

edwardneal replied:
That's fair. It's not ideal to see the existing serialization use 300-400 bytes of memory, then to see my changes double that usage on a buffer which isn't used - but the existing serialization has unmanaged memory usage behind it too, and by the time we've sent three controls the PR's memory usage should come out ahead.

I understand, and I'll change this (with the knock-on implication: those size calculations will also vanish.) Thanks for the explanations.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -651,10 +841,12 @@ public SortRequestControl(params SortKey[] sortKeys) : base("1.2.840.113556.1.4.
             for (int i = 0; i < sortKeys.Length; i++)
             {
                 _keys[i] = new SortKey(sortKeys[i].AttributeName, sortKeys[i].MatchingRule, sortKeys[i].ReverseOrder);
+                _keysAsnLength += 13 + (sortKeys[i].AttributeName?.Length ?? 0) + (sortKeys[i].MatchingRule?.Length ?? 0);
```
I could ask "what's 13?", then say "consider adding a comment", but my real answer is "stop doing manual size calculations".

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -699,101 +892,73 @@ public SortKey[] SortKeys
-                }
-            }
-            finally
-            {
-                if (control != IntPtr.Zero)
-                {
-                    LdapPal.FreeDirectoryControl(control);
-                }
+                            s_utf8Encoding.GetBytes(key.AttributeName, tmpValue);
+                            writer.WriteOctetString(tmpValue);
+                        }
+                        else
+                        {
+                            writer.WriteOctetString([]);
+                        }
```
This if-not-empty { count bytes; make space; UTF-8 encode; write octet string } else { write empty octet string } appears multiple times.  It should be factored out to `internal static void WriteLDAPString(AsnWriter writer, string value)`.  or something like that.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -836,8 +1004,7 @@ public VlvRequestControl(int beforeCount, int afterCount, string target) : this(
             if (target != null)
             {
-                UTF8Encoding encoder = new UTF8Encoding();
-                _target = encoder.GetBytes(target);
+                _target = s_utf8Encoding.GetBytes(target);
             }
```
This change makes invalid inputs (such as `"\uDD74\uD800"`) throw an `EncoderFallbackException`, which it used to not do.

I'm always a fan of rejecting nonsense inputs; but it's an observable change so it needs to have a test, and documented as a breaking change.

steveharter replied:
@edwardneal is there a test for the exception semantics above?

edwardneal replied:
I've added `VlvRequestControlTests.Ctor_InvalidUtf8Target_ThrowsEncoderFallbackException`.

bartonjs reviewed (commented):
@buyaa-n At this point I have completed my review of the technical impacts (that I noticed) of changing from BerConverter to AsnDecoder; with one outstanding issue regarding the encoding of SortResult/SortResponseControl's attribute value (https://github.com/dotnet/runtime/pull/101512/files?w=1#r1690564822).

I strongly oppose the current PR's efforts to do manual bookkeeping while trying to maintain multiple AsnWriter instances tiered by the size of their internal buffers; but it's really your call, not mine.

The test changes where patterns like `84 00 00 00 12` turn into `12` are all correct, those are both valid encodings (under BER) for the length 18 (0x12).

I am not personally confident that there's sufficient test coverage for this library to call this a "low risk" change.  I haven't done a thorough investigation of the coverage, only looked at what is in this diff.

The `DirectoryControl.cs` file really ought to be broken up into one-class-per-file... but definitely not concurrent with changes.

I am neither granting a checkmark nor mandating changes required.  I defer those decisions to the area owners.

edwardneal committed:
Further code review changes

* Refactored LDAP string writing logic into a separate WriteLdapString method in AsnWriterExtensions.
* Replaced separate AsnWriter instances with a single one (reduced memory usage when compared to multiple instances being active, even with varying buffer sizes.)
* Resolved backwards-compatibility issue when attempting to process an attribute name which contains invalid UTF-8 characters.
* Correcting expected ASN tag when parsing an attribute name in a sort result.

edwardneal commented:
I've resolved the outstanding issue with SortResponseControl, and addressed bartonjs' feedback by removing the tiered pooled AsnWriters.

> I am not personally confident that there's sufficient test coverage for this library to call this a "low risk" change. I haven't done a thorough investigation of the coverage, only looked at what is in this diff.

This comes at roughly the same point as I write the tests for the five DirectoryControls being modified, and their behavioural changes. Instead of including all of these tests in this PR, I'll extract the majority of those tests (including cases where BerConverter's parsing is lax/nonconformant to standard) into their own PR as a baseline against the existing method. I think that having this baseline should handle the direct risk, and we can come back to the current PR after the tests are verified passing on main.

buyaa-n commented:
> This comes at roughly the same point as I write the tests for the five DirectoryControls being modified, and their behavioural changes. Instead of including all of these tests in this PR, I'll extract the majority of those tests (including cases where BerConverter's parsing is lax/nonconformant to standard) into their own PR as a baseline against the existing method. I think that having this baseline should handle the direct risk, and we can come back to the current PR after the tests are verified passing on main.

@edwardneal this is what we wanted to have before merging/validating this PR. Glad to hear that you are planning to do it, now the main branch opened for `10.0` contribution we can unblock this PR when the tests for behavioral changes and edge case scenarios added

edwardneal commented:
Thanks @buyaa-n. I'd put the tests on the back burner, but with 10.0 now in progress I'll pick this back up and should hopefully have a set of tests to review within the next few days.

edwardneal committed:
Merge remote-tracking branch 'upstream/main' into issue-97540

edwardneal committed:
Update tests from PR107201

* Handles the switch to AsnDecoder
* One data correction in SortResponseControlTests - one field has a tag of [0] which BerConverter was ignoring

edwardneal committed:
Tightened validation based on updated tests

* AsqResponseControl now verifies that there's no trailing data within the control value's decoded BER sequence.
* All response controls now verify that there's no trailing data after their encoded value in the control value.

edwardneal committed:
Eliminated OS-specific test condition

Previously, a zero-length octet string interpreted via the "a" format string would have resulted in a null value in Windows 8.1, and an empty string in every other case. This now returns an empty string in all cases.

edwardneal commented:
Following the merge of #107201, I've updated this PR with three commits; it's ready for review at leisure.

#### Commits 1 & 3
* Updated the various response control tests, accounting for cases where previously "nonconformant but valid" samples are now marked as invalid, and eliminating the OS-specific behaviour in these tests. Commit 3 just removes the OS-specific behaviour from the conformant control samples - I'd forgotten that Windows 8.1 handled the "a" parameter differently.
* Corrected a handful of conformant values in the sort response control tests - I had specified the tag for an OCTET STRING (0x04), when the RFC actually gives the structure an explicit tag of `[0]` (0x80.)
* Clarified the differing behaviour of text decoding when the bytes to be decoded aren't a valid input to Encoding.GetString. .NET Framework/BerConverter throws a DecoderFallbackException, .NET 10 throws a BerConversionException.

#### Commit 2
* Slight tightening of the validation when comparing the nonconformant but valid samples between response tests: added a check to ensure that there's no trailing data inside the ASN sequence when parsing an AsqResponseControl.
* Also changed all five response control tests to make sure that there's no trailing data after the end of the RFC-compliant response control value. I think this tightens up the last piece of loose parsing of trailing data, and matches the RFCs/specs. I've tested this against an OpenLDAP server and can run the existing tests.

steveharter commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +145,174 @@ public virtual byte[] GetValue()
+                            }
+                            else
+                            {
+                                attributeNameBuffer = AsnDecoder.ReadOctetString(asnSpan, AsnEncodingRules.BER, out bytesConsumed);
```
Why is `SortResponseControl.AttributeNameTag` specified in above for `expectedTag ` in the call to `TryReadOctetString()` but not here in  `ReadOctetString()`?

Also there is no test coverage for the `else` here.

edwardneal replied:
Apologies - I'd missed this code path. I've added the test coverage, verified that it fails with the expected error and corrected the bug. There are no other paths with the same problem.

steveharter replied:
Thanks

steveharter commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -132,102 +145,174 @@ public virtual byte[] GetValue()
-                    object[] result = BerConverter.Decode("{iO}", value);
-                    Debug.Assert((result != null) && (result.Length == 2));
+                    Debug.Assert(controls[i] != null);
+                    byte[] value = controls[i]._directoryControlValue ?? Array.Empty<byte>();
```
When can `controls[i]._directoryControlValue` be null?

There is no test coverage for this branch.

steveharter replied:
I don't think we need a test if you removed the check - I was only verifying test coverage.

edwardneal replied:
Excellent - thanks for the review!

steveharter reviewed (commented):
@edwardneal I verified test coverage for the new code, and that looks good with the exception of the two cases I mentioned.

edwardneal committed:
Merge remote-tracking branch 'upstream/main' into issue-97540

edwardneal committed:
Code review: SortResponseControl

Reading a long attribute name would have failed due to an invalid expected ASN.1 tag. Correct, and added a test.

edwardneal committed:
Code review: VlvRequestControl

Added test to validate that passing an invalid UTF8 string as the target parameter of a VlvRequestControl will now throw an EncoderFallbackException.

edwardneal committed:
Code review: TransformControls

No longer null coalescing _directoryControlValue; replaced with a Debug.Assert that it's not null.

steveharter reviewed (approved):
Marking approved; will wait for a bit for any additional feedback from @bartonjs or others before merging. I think moving this code forward is a good thing even though there some risk. It also has the potential to fix issues including https://github.com/dotnet/runtime/issues/34679.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/AsnWriterExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+{
+    internal static class AsnWriterExtensions
+    {
+        public static void WriteLdapString(this AsnWriter writer, string value, Encoding stringEncoding, bool mandatory = true, Asn1Tag? tag = null)
```
I'm not sure this is a good name.  There's no construct (that I see) called `LdapString`, and not all strings in LDAP are sent as "A Utf8String, except using tag 04 instead of 0C".

`WriteUtf8OctetString`, maybe?

The `bool mandatory` has no peer on AsnWriter methods.  I recommend removing it here (making it always behave as `true`, and making the one "optional" caller bring that logic closer to home... so it looks like any other conditional write for an ASN OPTIONAL or DEFAULT value.

edwardneal replied:
The name `LdapString` partially comes from [RFC2251](https://datatracker.ietf.org/doc/html/rfc2251#section-4.1.5), as the backing type for `AttributeDescription`. Do you still want the name to change?

It was primarily used for writing the sort controls, and the other control logic piggybacks on the same method by explicitly specifying the encoding. I'll see if two methods would be clearer for this.

edwardneal replied:
I've tried a couple of different methods to see what the semantics look like, and agree - `WriteStringAsOctetString` it is. That's rolled up and done now.

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -87,10 +90,29 @@ public bool ReverseOrder
+        [ThreadStatic]
+        private static AsnWriter? t_writer;
+
+        [MemberNotNull(nameof(t_writer))]
```
I don't know how well `MemberNotNull` behaves with ThreadStatic.  No one should be touching `t_writer` except this function, so why is the annotation needed/warranted at all?

edwardneal replied:
Removed!

bartonjs commented on src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs:
```diff
@@ -87,10 +90,29 @@ public bool ReverseOrder
+
+        [MemberNotNull(nameof(t_writer))]
+        internal static AsnWriter GetWriter()
+            => t_writer ??= new AsnWriter(AsnEncodingRules.BER);
```
The correct behavior for every caller is to call `Reset()` on the writer when they get it, because they don't know if they have one that was abandoned due to an exception.

Maybe `GetWriter` should do that for them, instead?

edwardneal replied:
Thanks, I've shifted this around.

bartonjs commented:
> any additional feedback from @bartonjs

Had a few small things.  Looked at all the commits since [my last review](https://github.com/dotnet/runtime/pull/101512#pullrequestreview-2198032192).

edwardneal committed:
Disabled the new VlvRequestControl on .NET Framework

This change doesn't take effect on .NET Framework, so any test expecting an exception will fail.

edwardneal committed:
Cleanup GetWriter

Removed the unnecessary nullability annotations, and moved the Reset call into GetWriter.

edwardneal committed:
Renamed WriteLdapString to WriteStringAsOctetString

Also adjusted method signature to better align to the rest of the AsnWriter API surface.

steveharter commented:
@edwardneal do you have any further action items or planned changes? If not, I'll merge. Thanks.

edwardneal commented:
Thanks - I've responded to bartonjs' code review in-line, so don't have any further code changes planned.

An [earlier comment](https://github.com/dotnet/runtime/pull/101512#discussion_r1687254225) asked for a breaking change doc to be created though, and with the work settled I'll do this today/tomorrow.

edwardneal commented:
Thanks @steveharter and @bartonjs for your reviews. The breaking change doc is dotnet/docs#43885.

ericstj commented:
Removing `needs-breaking-change-doc-created` label as this PR already has a documentation issue: https://github.com/dotnet/docs/issues/43885
